File: spark/src/main/scala/com/lfw/spark/wc/WordCount.scala
package com.lfw.spark.wc

import org.apache.spark.rdd.RDD
import org.apache.spark.{SparkConf, SparkContext}

object WordCount {
  def main(args: Array[String]): Unit = {
    //TODO 1 创建SparkConf配置文件,并设置应用名称和运行模式
    val conf = new SparkConf().setAppName("WordCount").setMaster("local[*]") //可以交换顺序
    //TODO 2 构建SparkConf对象，作为编程入口
    val sc = new SparkContext(conf)
    //1 加载数据源
    val lineRDD: RDD[String] = sc.textFile("spark/input/wordcount.txt")
    //2 将一行一行的单词按照空格切割, 拍平 flatMap 算子安排
    //    val wordRDD: RDD[String] = lineRDD.flatMap(line=>line.split(" "))   完整版
    val wordRDD: RDD[String] = lineRDD.flatMap(_.split(" "))
    //3.将一个个的单词转换成(单词,1) map 算子安排
    //    val word2oneRDD: RDD(String,Int)] = wordRDD.map(word =>(word,1))
    val word2oneRDD: RDD[(String, Int)] = wordRDD.map((_, 1))
    //4 按照key分组 然后聚合进行规约操作 reduceByKey算子安排
    //    val word2sumRDD: RDD[(String, Int)] = word2oneRDD.reduceByKey((v1, v2) => (v1 + v2))
    val word2sumRDD: RDD[(String, Int)] = word2oneRDD.reduceByKey(_ + _)
    //将统计结果采集到控制台打印
    val result: Array[(String, Int)] = word2sumRDD.collect()
    // 触发执行 (行动算子)
    result.foreach(t => println(t))
    //TODO 3 关闭资源
    sc.stop()
  }
}

File: spark/src/main/java/com/lfw/spark/WordCount.java
package com.lfw.spark;

import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import org.apache.spark.api.java.function.FlatMapFunction;
import org.apache.spark.api.java.function.Function2;
import org.apache.spark.api.java.function.PairFunction;
import scala.Tuple2;

import java.util.Arrays;
import java.util.Iterator;

public class WordCount {
    public static void main(String[] args) {
        SparkConf sparkConf = new SparkConf().setAppName("JavaWordCount");
        //创建JavaSparkContext
        JavaSparkContext jsc = new JavaSparkContext(sparkConf);
        //使用 JavaSparkContext 创建RDD
        JavaRDD<String> lines = jsc.textFile(args[0]);

        JavaRDD<String> words = lines.flatMap(new FlatMapFunction<String, String>() {
            @Override
            public Iterator<String> call(String line) throws Exception {
                return Arrays.asList(line.split(" ")).iterator();
            }
        });

        //将单词组聚合
        JavaPairRDD<String, Integer> wordAndOne = words.mapToPair(
                new PairFunction<String, String, Integer>() {
                    @Override
                    public Tuple2<String, Integer> call(String word) throws Exception {
                        return Tuple2.apply(word, 1);
                    }
                }
        );
        //分组聚合
        JavaPairRDD<String, Integer> reduced = wordAndOne.reduceByKey(
                new Function2<Integer, Integer, Integer>() {
                    @Override
                    public Integer call(Integer v1, Integer v2) throws Exception {
                        return v1 + v2;
                    }
                }
        );
        //排序，先调换KV的顺序VK
        JavaPairRDD<Integer, String> swapped = reduced.mapToPair(
                new PairFunction<Tuple2<String, Integer>, Integer, String>() {
                    @Override
                    public Tuple2<Integer, String> call(Tuple2<String, Integer> tp) throws Exception {
                        return tp.swap();
                    }
                }
        );
        //再排序
        JavaPairRDD<Integer, String> sorted = swapped.sortByKey(false);
        //再调换顺序
        JavaPairRDD<String, Integer> result = sorted.mapToPair(
                new PairFunction<Tuple2<Integer, String>, String, Integer>() {
                    @Override
                    public Tuple2<String, Integer> call(Tuple2<Integer, String> tp) throws Exception {
                        return tp.swap();
                    }

                });
        //触发Action，将数据保存到HDFS
        result.saveAsTextFile(args[1]);
        //释放资源
        jsc.stop();
    }
}
